silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,393 @@
+        _storageDeposit(token, storageDepositAmount, getImplicitNearAccountIdForSelf(), false);
+    }
+
+    function recipientStorageDeposit(
+        IEvmErc20 token,
+        uint128 storageDepositAmount,
+        string memory recipientId
+    ) external whenNotPaused {
+        require(registeredRecipients[token][recipientId] == false, "The storage is already registered");
+        _storageDeposit(token, storageDepositAmount, recipientId, true);
+    }
+
+    function _storageDeposit(
+        IEvmErc20 token,
+        uint128 storageDepositAmount,
+        string memory recipientId,
+        bool isRecipientStorageDeposit
+    ) private {
+        string memory nearTokenAccountId = registeredTokens[token].nearTokenAccountId;
+        require(bytes(nearTokenAccountId).length > 0, "The token is not registered");
+
+        PromiseCreateArgs memory callStorageDeposit = near.call(
+            nearTokenAccountId,
+            "storage_deposit",
+            bytes(string.concat('{"account_id": "', recipientId, '", "registration_only": true }')),
+            storageDepositAmount,
+            BASE_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.storageDepositCallback.selector, token, recipientId, isRecipientStorageDeposit),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+        callStorageDeposit.then(callback).transact();
+    }
+
+    function storageDepositCallback(
+        IEvmErc20 token,
+        string memory recipientId,
+        bool isRecipientStorageDeposit
+    ) external onlyRole(CALLBACK_ROLE) {
+        require(
+            AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
+            "ERROR: The `storage_deposit()` XCC call failed"
+        );
+
+        if (isRecipientStorageDeposit) {
+            registeredRecipients[token][recipientId] = true;
+            emit RecipientStorageRegistered(token, recipientId);
+        } else {
+            TokenInfo storage tokenInfo = registeredTokens[token];
+            tokenInfo.isStorageRegistered = true;
+            emit TokenStorageRegistered(token, tokenInfo.nearTokenAccountId);
+        }
+    }
+
+    function ftTransferCallToNear(
+        IEvmErc20 token,
+        uint128 amount,
+        string calldata receiverId,
+        string calldata message
+    ) public payable whenNotPaused {
+        require(near.wNEAR.balanceOf(address(this)) >= ONE_YOCTO, "Not enough wNEAR balance");
+        TokenInfo memory tokenInfo = registeredTokens[token];
+        require(tokenInfo.isStorageRegistered, "The token storage is not registered");
+
+        if (address(token) == address(0)) {
+            require(msg.value == amount, "Incorrect attached value");
+            _withdrawNativeTokenToNear(bytes(getImplicitNearAccountIdForSelf()), msg.value);
+        } else {
+            token.transferFrom(msg.sender, address(this), amount);
```

Comment:
Should we also check that `msg.value == 0` in that case?

Reply:
d4d939d2dbc2c00fba2ced5a1ac656877c47a321